include/rmm/prefetch.hpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/error.hpp>
+
+#include <cuda/std/span>
+
+namespace rmm {
+
+/**
+ * @addtogroup utilities
+ * @{
+ * @file
+ */
+
+/**
+ * @brief Prefetch memory to the specified device on the specified stream.
+ *
+ * This function is a no-op if the pointer is not to CUDA managed memory.
+ *
+ * @throw rmm::cuda_error if the prefetch fails.
+ *
+ * @tparam T The type of the elements pointed to by `ptr`.
+ * @param ptr The pointer to the memory to prefetch
+ * @param size The number of bytes to prefetch
+ * @param device The device to prefetch to
+ * @param stream The stream to use for the prefetch
+ */
+template <typename T>
+void prefetch(T* ptr, std::size_t size, rmm::cuda_device_id device, rmm::cuda_stream_view stream)
+{
+  auto result = cudaMemPrefetchAsync(ptr, size, device.value(), stream.value());
+  // InvalidValue error is raised when non-managed memory is passed to cudaMemPrefetchAsync
+  // We should treat this as a no-op
+  if (result != cudaErrorInvalidValue && result != cudaSuccess) { RMM_CUDA_TRY(result); }
+}
+
+/**
+ * @brief Prefetch a span of memory to the specified device on the specified stream.
+ *
+ * This function is a no-op if the buffer is not backed by CUDA managed memory.
+ *
+ * @throw rmm::cuda_error if the prefetch fails.
+ *
+ * @param data The span to prefetch
+ * @param device The device to prefetch to
+ * @param stream The stream to use for the prefetch
+ */
+template <typename T>
```

Comment:
*note (nothing to do here)*: A slight annoyance that without concepts I think there's no way to avoid writing the template parameter at the call site, but if we had C++20 then:
```c++
template <typename T>
requires requires(T&& s) { cuda::std::span(std::forward<T>(s)); }
void prefetch(T&& data, rmm::cuda_device_id device, rmm::cuda_stream_view stream)
{
  cuda::std::span sp(std::forward<T>(data));
  prefetch(sp.data(), sp.size_bytes(), device, stream);
}
```
Would allow one to write:
```c++
rmm::device_vector<int> buf{...};
prefetch(buf, ...);
```
Without the need for the template parameter to `prefetch`, I think.

Reply:
Yes, the current required interface to support `span` is annoying.

Reply:
I need to learn more about concepts to understand what your code is saying.

Reply:
> doesn't calling this function with an rmm::device_buffer automatically convert to a span? 

Yes, but template argument deduction doesn't work. So you have to write:

```c++
rmm::device_buffer<int> buf{...};
prefetch<int>(buf, ...);
```

Rather than what you feel like should work:
```c++
rmm::device_buffer<int> buf{...};
prefetch(buf, ...);
```

And yes, that requires concept is basically the concepts version of your SFINAE constraint. It says, I'll take any && reference as long as it is span constructible.

Reply:
I mistyped, sub device_vector<int> for device_buffer